### PR TITLE
Enhancement/field line traces

### PR DESCRIPTION
**Pull request by gmalinve**
- Fix in create_fieldplot for lines
- Implement field plot line traces, creation and update
- field plot line traces test

### CODE REVIEW

**Review by maxcapodi78 — APPROVED** [review]

File: pyaedt/modules/PostProcessor.py [hunk 1]
```diff
@@ -2824,6 +2873,106 @@ def create_fieldplot_line(self, objlist, quantityName, setup_name=None, intrinsi
...
+        surface_tracing_objs : list
+            List of the surface tracing objects.
+        setup_name : str, optional
+            Name of the setup in the format ``"setupName : sweepName"``. The default
+            is ``None``.
+        intrinsincDict : dict, optional
+            Dictionary containing all intrinsic variables. The default
+            is ``{}``.
+        plot_name : str, optional
+            Name of the fieldplot to create.
```
Comment (MaxJPRey) [inline pyaedt/modules/PostProcessor.py:2904]:
```suggestion
            Name of the fieldplot to create. The default is ``None``.
```

File: pyaedt/modules/PostProcessor.py [hunk 2]
```diff
@@ -2788,6 +2788,55 @@ def _create_fieldplot(self, objlist, quantityName, setup_name, intrinsics, listt
...
+            self.ofieldsreporter.SetPlotsViewSolutionContext([plot_name], setup_name, "Time:" + intrinsics["Time"])
+        if plt:
+            self.field_plots[plot_name] = plot
+            return plot
+        else:
+            return False
+
     @pyaedt_function_handler()
     def create_fieldplot_line(self, objlist, quantityName, setup_name=None, intrinsincDict=None, plot_name=None):
         """Create a field plot of line.
```
Comment (PipKat) [inline pyaedt/modules/PostProcessor.py:2842]:
```suggestion
        """Create a field plot of the line.
```

File: pyaedt/modules/PostProcessor.py [hunk 3]
```diff
@@ -2824,6 +2873,106 @@ def create_fieldplot_line(self, objlist, quantityName, setup_name=None, intrinsi
...
+        self,
+        seeding_faces,
+        in_volume_tracing_objs=None,
+        surface_tracing_objs=None,
+        setup_name=None,
+        intrinsincDict=None,
+        plot_name=None,
+    ):
+        """
+        Create a field plot of line.
```
Comment (PipKat) [inline pyaedt/modules/PostProcessor.py:2887]:
```suggestion
        Create a field plot of the line.
```

File: pyaedt/modules/PostProcessor.py [hunk 4]
```diff
@@ -2824,6 +2873,106 @@ def create_fieldplot_line(self, objlist, quantityName, setup_name=None, intrinsi
...
+
+        Parameters
+        ----------
+        seeding_faces : list
+            List of seeding faces.
+        in_volume_tracing_objs : list
+            List of the in volume tracing objects.
```
Comment (PipKat) [inline pyaedt/modules/PostProcessor.py:2894]:
```suggestion
            List of the in-volume tracing objects.
```

File: pyaedt/modules/PostProcessor.py [hunk 5]
```diff
@@ -2824,6 +2873,106 @@ def create_fieldplot_line(self, objlist, quantityName, setup_name=None, intrinsi
...
+            self.logger.info("Plot {} exists. returning the object.".format(plot_name))
+            return self.field_plots[plot_name]
+        if not isinstance(seeding_faces, list):
+            seeding_faces = [seeding_faces]
+        seeding_faces_ids = []
+        for face in seeding_faces:
+            if self._app.modeler[face]:
+                seeding_faces_ids.append(self._app.modeler[face].id)
+            else:
+                self.logger.error("Object {} doesn't exists in current design".format(face))
```
Comment (PipKat) [inline pyaedt/modules/PostProcessor.py:2931]:
```suggestion
                self.logger.error("Object {} doesn't exist in current design".format(face))
```

File: pyaedt/modules/PostProcessor.py [hunk 6]
```diff
@@ -2824,6 +2873,106 @@ def create_fieldplot_line(self, objlist, quantityName, setup_name=None, intrinsi
...
+        if not in_volume_tracing_objs:
+            in_volume_tracing_ids = [0]
+        elif not isinstance(in_volume_tracing_objs, list):
+            in_volume_tracing_objs = [in_volume_tracing_objs]
+            in_volume_tracing_ids = []
+            for obj in in_volume_tracing_objs:
+                if self._app.modeler[obj]:
+                    in_volume_tracing_ids.append(self._app.modeler[obj].id)
+                else:
+                    self.logger.error("Object {} doesn't exists in current design".format(obj))
```
Comment (PipKat) [inline pyaedt/modules/PostProcessor.py:2942]:
```suggestion
                    self.logger.error("Object {} doesn't exist in current design".format(obj))
```

File: pyaedt/modules/PostProcessor.py [hunk 7]
```diff
@@ -2824,6 +2873,106 @@ def create_fieldplot_line(self, objlist, quantityName, setup_name=None, intrinsi
...
+                    return False
+        else:
+            self.logger.error("Provide a valid object for in volume objects selection.")
```
Comment (PipKat) [inline pyaedt/modules/PostProcessor.py:2945]:
```suggestion
            self.logger.error("Provide a valid object for the in-volume objects selection.")
```
Can objects be singular in the message: Provide a valid object for the in-volume object selection."

File: pyaedt/modules/solutions.py [hunk 8]
```diff
@@ -2786,6 +2872,22 @@ def field_plot_settings(self):
             ],
         ]
 
+    @property
+    def field_line_trace_plot_settings(self):
+        """Field line trace plot settings.
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:2877]: Can this description (and the description for what is returned) be rewritten. It has four adjectives in front of the noun "settings".  Maybe "Settings for the trace plot of the field line."

File: pyaedt/modules/solutions.py [hunk 9]
```diff
@@ -2813,12 +2920,52 @@ def update(self):
         bool
             ``True`` when successful, ``False`` when failed.
         """
-        self.oField.ModifyFieldPlot(self.name, self.surfacePlotInstruction)
+        try:
+            if self.seeding_faces:
+                if self.seeding_faces[0] != len(self.seeding_faces) - 1:
+                    for face in self.seeding_faces[1:]:
+                        if not isinstance(face, int):
+                            self._postprocessor.logger.error("Provide the object valid id.")
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:2928]: This message seems odd. Maybe "Provide the valid ID for the object."

File: pyaedt/modules/solutions.py [hunk 10]
```diff
@@ -2813,12 +2920,52 @@ def update(self):
...
+                            return False
+                        else:
+                            if face not in list(self._postprocessor._app.modeler.objects.keys()):
+                                self._postprocessor.logger.error("Invalid object id.")
+                                return False
+                    self.seeding_faces[0] = len(self.seeding_faces) - 1
+                if self.volume_indexes[0] != len(self.volume_indexes) - 1:
+                    for obj in self.volume_indexes[1:]:
+                        if not isinstance(obj, int):
+                            self._postprocessor.logger.error("Provide the object valid id.")
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:2938]: Same comment as earlier.

**Review by maxcapodi78 — COMMENTED** [review]
Please improve Coverage and fix comment suggestions

File: pyaedt/modules/solutions.py [hunk 11]
```diff
@@ -2813,12 +2920,52 @@ def update(self):
...
+            return True
+        except:
+            return False
 
     @pyaedt_function_handler()
     def update_field_plot_settings(self):
         """Modify the field plot settings.
 
+        .. note::
+            Not available for field plot line traces.
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:2967]:
```suggestion
            This method is not available for field plot line traces.
```
The note must be a complete sentence. Again, there are lots of adjectives before traces. Maybe it's not a comprehension problem for your audience.

File: pyaedt/modules/solutions.py (hunk 8 above)
Comment (gmalinve) [reply pyaedt/modules/solutions.py:2877]: sure! thanks for feedback!

File: pyaedt/modules/solutions.py [hunk 12]
```diff
@@ -2786,6 +2872,22 @@ def field_plot_settings(self):
...
         ]
 
+    @property
+    def field_line_trace_plot_settings(self):
+        """Settings for the field line traces in the plot.
+
+        Returns
+        -------
+        list
+            Field line trace plot settings.
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:2882]:
```suggestion
            List of settings for the field line traces in the plot.
```
Thanks for making changes to eliminate that long string of adjectives!

File: pyaedt/modules/solutions.py [hunk 13]
```diff
@@ -2813,12 +2920,52 @@ def update(self):
...
+                            return False
+                        else:
+                            if face not in list(self._postprocessor._app.modeler.objects.keys()):
+                                self._postprocessor.logger.error("Invalid object id.")
+                                return False
+                    self.seeding_faces[0] = len(self.seeding_faces) - 1
+                if self.volume_indexes[0] != len(self.volume_indexes) - 1:
+                    for obj in self.volume_indexes[1:]:
+                        if not isinstance(obj, int):
+                            self._postprocessor.logger.error("Provide valid object id for in volume object.")
```
Comment (PipKat) [inline pyaedt/modules/solutions.py:2938]:
```suggestion
                            self._postprocessor.logger.error("Provide valid object id for in-volume object.")
```

File: pyaedt/modules/PostProcessor.py [hunk 14]
```diff
@@ -2824,6 +2873,106 @@ def create_fieldplot_line(self, objlist, quantityName, setup_name=None, intrinsi
...
+        seeding_faces : list
+            List of seeding faces.
+        in_volume_tracing_objs : list
+            List of the in-volume tracing objects.
+        surface_tracing_objs : list
+            List of the surface tracing objects.
+        setup_name : str, optional
+            Name of the setup in the format ``"setupName : sweepName"``. The default
+            is ``None``.
+        intrinsincDict : dict, optional
```
Comment (MaxJPRey) [inline pyaedt/modules/PostProcessor.py:2900]:
```suggestion
        intrinsinc_dict : dict, optional
```

File: pyaedt/modules/PostProcessor.py [hunk 15]
```diff
@@ -2824,6 +2873,106 @@ def create_fieldplot_line(self, objlist, quantityName, setup_name=None, intrinsi
...
+            in_volume_tracing_ids.insert(0, len(in_volume_tracing_ids))
+        if surface_tracing_ids != [0]:
+            surface_tracing_ids.insert(0, len(surface_tracing_ids))
+        return self._create_fieldplot_line_traces(
+            seeding_faces_ids,
+            in_volume_tracing_ids,
+            surface_tracing_ids,
+            "FieldLineTrace",
+            setup_name,
+            intrinsincDict,
```
Comment (MaxJPRey) [inline pyaedt/modules/PostProcessor.py:2972]:
```suggestion
            intrinsinc_dict,
```

**Review by MaxJPRey — APPROVED** [review]

**Pull request merged**